aggregator/src/aggregator/aggregation_job_continue.rs

```diff
@@ -117,84 +116,91 @@ impl VdafOps {
                 }
             };
 
-            // Parse preparation message out of prepare step received from leader.
-            let prep_msg = match prep_step.result() {
-                PrepareStepResult::Continued(payload) => A::PrepareMessage::decode_with_param(
-                    prep_state,
-                    &mut Cursor::new(payload.as_ref()),
-                )?,
-                _ => {
-                    return Err(datastore::Error::User(
-                        Error::UnrecognizedMessage(
-                            Some(*task.id()),
-                            "leader sent non-Continued prepare step",
+            let (report_aggregation_state, prepare_step_result, output_share) =
+                trace_span!("VDAF preparation")
+                    .in_scope(|| {
+                        // Evaluate the stored transition to recover our current state.
+                        transition
+                            .evaluate(vdaf.as_ref())
```

Comment:
Oh, I see: we evaluate twice to get the result of the next transition, but still store `new_transition` (which is the correct thing to store). Maybe we can factor this to store the correct data without needing to call `prep_next` twice by calling `evaluate` twice? I think we'd need to store `new_state` rather than `new_transition`, but the state is what we're getting from the doubled-up call to `evaluate`.

Reply:
In this case we are trading off compute time against storage space. Let's consider the case where we are in step _n_, then call `evaluate` and transition to `Waiting`. The outputs will be:

- state for round _n+1_,
- prepare message from round _n_,
- prepare message share from round _n+1_.

We transmit the prepare message and the prepare message share to the helper (this is in a `PingPongMessage`). Now, if we want to avoid calling `evaluate` twice, then we have to store the round _n+1_ prepare state and the round _n+1_ prepare message share. Prepare message shares are large so we don't want to do that.

So instead, after sending the message, we serialize the `Transition`, and then we only have to store:

- state for round _n_
- prepare message from round _n_

With these two things, we can recompute the state for round _n+1_ and the prepare message share from round _n+1_, at the cost of re-running `Transition::evaluate`.

All that being said, I'm still thinking about ways to refactor `Transition::evaluate` so that it can consume `self`. I think there may be a way to do this that avoids doing computations twice.

Reply:
Having studied this some more, I'm not sure we can do better on this tradeoff. I went back and looked at your prototype of ping-pong (#1234) and I notice that when its helper goes into Waiting, it will store:

- state for round _n_
- prepare message for round _n - 1_
- prepare message share from round _n_

([source](https://github.com/divviup/janus/blob/160d698db871e96f3d545c8240ddbe62671523e0/aggregator/src/aggregator.rs#L1336))

Over in the [libprio change](https://github.com/divviup/libprio-rs/pull/683#pullrequestreview-1584955757) we discussed how we don't want to store the large prepare message shares, so I introduced `Transition` to alleviate storage requirements at the cost of more compute time because we re-evaluate the transition.

The only other thing I can think to do would be to store two prepare states when the helper is waiting. Then, if it needs to recompute a previously transmitted prep share, it uses the older state. If it needs to advance to the next round, it uses the newer state. But I think that gets quite complicated, and for many VDAFs, storing two prepare states is about as bad as storing a prepare share.

I'm open to doing better here but I don't really see how.

Reply:
The report aggregation state stored by #1234 includes only the state: https://github.com/divviup/janus/blob/160d698db871e96f3d545c8240ddbe62671523e0/aggregator/src/aggregator.rs#L1384

(the last prep step portion of the report aggregation also incldues a prep message & prep message share for idempotency/replay, but this is also stored by this PR, so that part is the same for both implementations)

This PR stores the transition, which is a [prepare state and a message](https://github.com/divviup/libprio-rs/blob/10598564e7256b135aec0a978c6901583491b926/src/topology/ping_pong.rs#L191-L198); for the last prep step storage it additionally stores a prep message & prep share (same as in #1234, as noted above). So by storing the transition, we are storing an additional prepare message, as well as evaluating the transition twice.

I think we should aim for an abstraction that does not cause the Helper to store additional data, i.e. the report aggregation state (modulo the last prep step) stores only the preparation state. I think this would also eliminate the need to evaluate each transition twice.

Reply:
I had to think about this for a while (actually, what I really needed was a nap) but I get what you mean now. I went with (a): if `Transition::evaluate` returns `(PingPongState::Continued(prep_state), message)`, the helper now stores `prep_state` in its `ReportAggregation` (the leader still stores the `Transition`).

One awkward thing here is that we now have `ReportAggregationState::WaitingLeader(PingPongTransition)` and `ReportAggregationState::WaitingHelper(A::PrepareState)`. The nice thing is that this is expressive and the compiler can help us ensure we pass in appropriate values. The downside is that we have to handle one extra enum variant everywhere we work with `ReportAggregationState`.